pkg/generator/specification.go

```diff
@@ -113,56 +122,100 @@ func (s *SpecificationGenerator) generateIATPerGranularity(numberOfInvocations i
 				break
 			}
 		}
-		beginningIAT := sum - split
-		endIAT := iatResult[i] - beginningIAT
-		finalIAT := append([]float64{beginningIAT}, iatResult[i+1:]...)
-		finalIAT = append(finalIAT, iatResult[:i]...)
-		iatResult = append(finalIAT, endIAT)
-	} else {
-		iatResult = append([]float64{0.0}, iatResult...)
+
+		if i < len(iatResult) && i+1 < len(iatResult) {
+			beginningIAT := sum - split
+			endIAT := iatResult[i] - beginningIAT
+			finalIAT := append([]float64{beginningIAT}, iatResult[i+1:]...)
+			finalIAT = append(finalIAT, iatResult[:i]...)
+			iatResult = append(finalIAT, endIAT)
+		}
 	}
 
 	return iatResult, totalDuration
 }
 
+func getBlankTimeUnit(granularity common.TraceGranularity) float64 {
+	if granularity == common.MinuteGranularity {
+		return 60_000_000
+	} else {
+		return 1_000_000
+	}
+}
+
 // GenerateIAT generates IAT according to the given distribution. Number of minutes is the length of invocationsPerMinute array
-func (s *SpecificationGenerator) generateIAT(invocationsPerMinute []int, iatDistribution common.IatDistribution, shiftIAT bool, granularity common.TraceGranularity) (common.IATMatrix, common.ProbabilisticDuration) {
-	var IAT [][]float64
+func (s *SpecificationGenerator) generateIAT(invocationsPerMinute []int, iatDistribution common.IatDistribution,
+	shiftIAT bool, granularity common.TraceGranularity) (common.IATArray, []int, common.ProbabilisticDuration) {
+
+	var IAT []float64
+	var perMinuteCount []int
 	var nonScaledDuration []float64
 
+	emptyTime := 0.0
+	firstInvocation := true
+	lastNonZeroIndex := -1
+
 	numberOfMinutes := len(invocationsPerMinute)
 	for i := 0; i < numberOfMinutes; i++ {
 		minuteIAT, duration := s.generateIATPerGranularity(invocationsPerMinute[i], iatDistribution, shiftIAT, granularity)
 
-		IAT = append(IAT, minuteIAT)
+		perMinuteCount = append(perMinuteCount, len(minuteIAT))
+		// for distribution tests in Go unit tests
 		nonScaledDuration = append(nonScaledDuration, duration)
+
+		if perMinuteCount[i] == 0 {
+			// accumulate empty time
+			emptyTime += getBlankTimeUnit(granularity)
+			continue
+		} else {
+			// if perMinuteCount for last non-empty minute == 1
+			// then X = getBlankTimeUnit(granularity)	=> case when minute was not split
+			// else X = IAT[len(IAT) - 1]				=> case when minute was split
+			// minuteIAT[0] = emptyTime + X
+
+			minuteIAT[0] += emptyTime
+			if !firstInvocation && perMinuteCount[lastNonZeroIndex] != 1 {
+				minuteIAT[0] += IAT[len(IAT)-1]
```

Comment:
This doesn't seem right. This works exclusively for the equidistant now, it won't work for any other IAT.

You changed the generateIATPerGranularity in a way that it doesn't add up to a full minute (you can see that in tests that you added). So, the last value in minuteIAT doesn't correspond to the interval after the last invocation to the end of minute, but interval between two last invocations. In case of equidistant, they are the same, so the tests work fine. But there are no tests for other IATs with several minutes of invocations.

I propose you return the last element of the array (essentially, just revert the changes to generateIATPerGranularity), and use `IAT[len(IAT)-1]` as the `emptyTime`, as I outlined in previous suggestion yesterday.

Reply:
Fixed in https://github.com/vhive-serverless/invitro/pull/557/commits/134b82c6b3a21b974bf84c914b2668cfe2968ff8.